### PR TITLE
Automatic Quote selection: Use priorities for price/market selection

### LINKED ISSUE #1273: Smart selection of quote token
Smart selection of quote token and base token

Rafa's suggestion for simple and good enough solution is:
- We need to know when a token is a stable coin
- If it's stable coin vs non stable coin, we choose stable coin as a quote
- In any other case, we continue as now: sell is the quote

With the above, we can eliminate the need of showing the 2 prices.
We can, just in case allow to invert the price anyways.

_Estimation: 3.5_

### PR DESCRIPTION
**Pull request by W3stside**
Closes #1273 

Open and supersedes #1344 

Tests: `yarn test test/utils/adjustMarketPrice.spec.ts test/hooks/quoteTokenPriorities.hooks.test.tsx`
- Runs both the util function that adjusts token priorities and the hook that returns tokenList with priorities

There was difficulty getting the correct prices displayed considering how many places price is manipulated, e.g:
1. `PriceEstimations` calculates price based on base/quote token, but this led to price errors when the priority algorithm and hook would change the order, resulting in inverse price
2. `Price` the inputs are conditionally shown based on `inversePrice` and show price accordingly. This is tricky if the market is switched due to high priority market pair being traded, thus re-inversing the price.
3. SwapPrice and the inversing of markets with adjusted pairs was similar to point 3

### CODE REVIEW

**Comment by W3stside** [comment]
@anxolin I believe I addressed all your comments from the previous, closed PR, #1344

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr1350--dexreact.review.gnosisdev.com

**Comment by W3stside** [comment]
Still a WIP as price is not shown properly and there is test code lingering...

File: src/components/TradeWidget/index.tsx [hunk 1]
```diff
@@ -543,8 +558,7 @@ const TradeWidget: React.FC = () => {
   const defaultValidFrom = trade.validFrom || validFromParam
   const defaultValidUntil = trade.validUntil || validUntilParam
 
-  const [priceShown, setPriceShown] = useState<'INVERSE' | 'DIRECT'>('INVERSE')
-
+  const [priceShown, setPriceShown] = useState<'INVERSE' | 'DIRECT'>('DIRECT')
```
Comment (anxolin) [inline src/components/TradeWidget/index.tsx:561]: ubernit: and also not from this PR, but can we just give the type a name (create the type)
i.e.
```ts
type PriceRepresentation =  'INVERSE' | 'DIRECT'
```

File: src/hooks/usePrioritiseTokensForPrice.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+export function checkMarketAndSmartAdjust({ sellToken, receiveToken }: HookProps): SmartTokenStruct {
+  let baseToken = receiveToken
+  let quoteToken = sellToken
+  let wasPriorityAdjusted = false
+
+  // Prio: 1    - STABLE USD
+  // Prio: 2    - STABLE EUR
+  // Prio: 3    - WETH
+  // Prio: 99   - VOLATILE
+  const basePrio = baseToken.priority || 99
```
Comment (anxolin) [inline src/hooks/usePrioritiseTokensForPrice.ts:32]: 99 is arbitrary, I would use `Number.MAX_SAFE_INTEGER`

Comment (anxolin) [reply src/hooks/usePrioritiseTokensForPrice.ts:32]: Also, if you do a small function getPriority, you can refactor it
```js
const basePrio = getPriority(baseToken)
const quotePrio = getPriority(quotePrio)
```

Also, I wouldn't shorten `priority` word

File: test/utils/adjustMarketPrice.spec.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+// QUOTE gives price
+//     BASE/QUOTE
+// e.g WETH/TUSD
+
+describe('Check market and check base/quote token priorities', () => {
+  const [WETH, TUSD, USDT, , , , , FAKE] = tokenList
+
+  test('STABLE/STABLE => return with no change', () => {
+    // QUOTE
+    const sellToken = TUSD
```
Comment (anxolin) [inline test/utils/adjustMarketPrice.spec.ts:13]: Why don't you define the constants for the token `USDT`, `TUSD` , so they have already the priority (instead of change in it in every test)

I think, if you do this, the tests become a bit shorter, easier to read, but more importantly less error prone, since you cannot mess up with priorities. 
We can use prio 1 for TUSD and USDT, 2 for EURO, 3 for WETH, ... (and do not change it)

Also, you can refactor the test to be onliners if you build a function:

```ts
 assertMarket({ TUSD, USDT  },  {  baseToken: TUSD, quote: USDT }) 
```

File: src/components/TradeWidget/Price.tsx [hunk 4]
```diff
@@ -225,10 +227,10 @@ const Price: React.FC<Props> = ({
   return (
     <Wrapper>
       <strong>
-        Limit Price <OrderBookBtn baseToken={receiveToken} quoteToken={sellToken} />
+        Limit Price <OrderBookBtn baseToken={baseToken} quoteToken={quoteToken} />
       </strong>
       {/* using display: none to hide to avoid hook-form reregister */}
-      <PriceInputBox hidden={priceShown !== 'DIRECT'}>
+      <PriceInputBox hidden={(isPriceInverted && wasPriorityAdjusted) || (!isPriceInverted && !wasPriorityAdjusted)}>
```
Comment (anxolin) [inline src/components/TradeWidget/Price.tsx:233]: I think the logic is a bit complex, instead of complicating when rendering the view, can we just use "aux" variables for the quote/base returned by the function, and make the final selection of quote/base based on that plus `isPriceInverted`. 

This way, you move the logic to the JS part

File: src/components/TradeWidget/Price.tsx [hunk 5]
```diff
@@ -156,13 +156,14 @@ export const PriceInputBox = styled.div<{ hidden?: boolean }>`
...
-  receiveToken: TokenDetails
+  quoteToken: TokenDetails
+  baseToken: TokenDetails
   priceInputId: string
   priceInverseInputId: string
   tabIndex?: number
   swapPrices: () => void
-  priceShown: 'INVERSE' | 'DIRECT'
+  isPriceInverted: boolean
+  wasPriorityAdjusted: boolean
```
Comment (anxolin) [inline src/components/TradeWidget/Price.tsx:166]: Why do we need the `wasPriorityAdjusted`  why is it relevant?
I mean, the important thing is what's the new quote, not if it's the one we thought originally, or not. Probably I'm missing sth

File: test/utils/adjustMarketPrice.spec.ts [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+import tokenList from '../data/tokenList'
+
+// QUOTE gives price
+//     BASE/QUOTE
+// e.g WETH/TUSD
+
+describe('Check market and check base/quote token priorities', () => {
+  const [WETH, TUSD, USDT, , , , , FAKE] = tokenList
+
+  test('STABLE/STABLE => return with no change', () => {
```
Comment (anxolin) [inline test/utils/adjustMarketPrice.spec.ts:11]: Ubernit: you can create inner "describe" and group by "STABLE/STABLE", then you can do the inner tests for this two.

USDT/USDC  --> USDT/USDC
USDC/USDT  --> USDC/USDT

not ubernit: I think you haven't tested the second case above ^^^.

**Review by anxolin — COMMENTED** [review]
That's much better!
And I really like the tests

File: src/components/TradeWidget/Price.tsx (hunk 5 above)
Comment (W3stside) [reply src/components/TradeWidget/Price.tsx:166]: BASE_TOKEN/QUOTE_TOKEN: `DAI/WETH`

`DAI` = BASE TOKEN
`WETH` = QUOTE TOKEN

`DAI` PRIORITY = 1
`WETH` PRIORITY = 3

SHOW MARKET AS: `WETH/DAI` i.e
`DAI` = QUOTE TOKEN
`WETH = BASE TOKEN

In this scenario, `PriceSuggestions` receive the `baseToken` and `quoteToken` props as:
(they were switched above, according to the priority scheme)
```
<PriceSuggestions
  ...
  baseToken={WETH}
  quoteToken={DAI}
  isPriceInverted={priceInvertedState}
/>

where, inside `PriceSuggestions` - the `usePriceWithSlippage` hook is run using the `baseToken` `quoteToken` specified props above

So the price using the adjusted tokens will return the `inversePrice` essentially as the token pair has been switched, meaning we need some way of knowing when it has been switched to adjust the pricing shown

Open to any easier approaches. One thing to keep in mind while testing is the user's choice to `inversePrice` and how to accurately portray that in the multiple places we show the inverse price button and the `Receive at least` + `Limit Price` inputs

File: test/utils/adjustMarketPrice.spec.ts (hunk 3 above)
Comment (W3stside) [reply test/utils/adjustMarketPrice.spec.ts:13]: i like doing extra work, obviously

File: src/components/TradeWidget/index.tsx [hunk 7]
```diff
@@ -394,6 +395,20 @@ export interface TradeFormData {
 
 const validationResolver = resolverFactory<TradeFormData>(validationSchema)
 
+// TESTING
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/components/TradeWidget/index.tsx:398]: nope

**Comment by anxolin** [comment]
At least in Rinkeby, the selection is not correct, look at this picture:
image

File: src/components/TradeWidget/Price.tsx (hunk 5 above)
Comment (anxolin) [reply src/components/TradeWidget/Price.tsx:166]: I understand why `isPriceInverted` is needed, but not `wasPriorityAdjusted`.

If you hand the final base/quote and the isPriceInverted, any component should be able to represent it.
Also, most components wouldn't even need the "isPriceInverted" probably. I can see why for price suggestions does.

> Open to any easier approaches

One idea is modeling as having JUST `sellToken`, `buyToken`,  `baseToken` and `quoteToken`.
the two first is what you have selected in the form
the two last define the market. Just thought this might model more what we have in there.

**Comment by anxolin** [comment]
By the way, the lint issues shouldn't be fixed in this PR. It's not feasible to review a PR with code changes and 144 changes

File: src/data/quoteTokenPriorities.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      ],
+    },
+  },
+  {
+    // OWL
+    priority: 2,
+    addresses: {
+      [Network.Mainnet]: ['0x1a5f9352af8af974bfc03399e3767df6370d82e4'], // OWL
+      // Rinkeby
+      [Network.Rinkeby]: [],
```
Comment (anxolin) [inline src/data/quoteTokenPriorities.ts:51]: You didn't add OWL in Rinkeby for any particular reason?

File: src/hooks/usePrioritiseTokensForPrice.ts (hunk 2 above)
Comment (anxolin) [reply src/hooks/usePrioritiseTokensForPrice.ts:32]: You resolved my comment but I still see the same issue...

**Review by anxolin — COMMENTED** [review]
You still have some unaddressed comments.

I have still those same questions as last review. Please, make sure you address or reply before re-requesting review. If if you think is easier to discuss in a zoom, that's an option too.

**Comment by W3stside** [comment]
@anxolin ah yes I changed the `Number.MAX_INTEGER` into the base priority setting area - `useTokenList`

```
const map: Map<string, number> = new Map()
  quoteTokenPrioritiesList.forEach(({ priority, addresses }) => {
    if (addresses[networkId] && addresses[networkId].length) {
      addresses[networkId].forEach((address: string) =>
        map.set(address.toLowerCase(), priority || Number.MAX_SAFE_INTEGER),
      )
    }
  })
```

File: src/data/quoteTokenPriorities.ts (hunk 8 above)
Comment (W3stside) [reply src/data/quoteTokenPriorities.ts:51]: nope, guess it slipped my mind :+1:

File: src/hooks/usePrioritiseTokensForPrice.ts (hunk 2 above)
Comment (W3stside) [reply src/hooks/usePrioritiseTokensForPrice.ts:32]: it's in a different section, `useTokenList` but i will remove this as it's no longer needed

File: src/components/TradeWidget/index.tsx (hunk 1 above)
Comment (anxolin) [reply src/components/TradeWidget/index.tsx:228]: given that's a easy change, could you consider doing this?

**Review by anxolin — COMMENTED** [review]
It looks broken

image

image

File: src/hooks/useTokenList.ts [hunk 9]
```diff
@@ -3,21 +3,66 @@ import { TokenDetails } from 'types'
...
+  quoteTokenPrioritiesList?: typeof quoteTokenPriorities
+}): TokenDetails[] {
+  if (!networkId) return tokenList
+  const stablecoinMap = getNetworkCorrectStableCoinList(networkId, quoteTokenPrioritiesList)
+  console.debug('stablecoinMap', stablecoinMap)
+  // no networkId or map is empty as network id returns no tokens
+  if (!stablecoinMap?.size) return tokenList
+
+  return tokenList.map<TokenDetails>((token) =>
+    Object.assign({}, token, { priority: stablecoinMap.get(token.address.toLowerCase()) || Number.MAX_SAFE_INTEGER }),
```
Comment (Velenir) [inline src/hooks/useTokenList.ts:46]:
```suggestion
    ( {...token, priority: stablecoinMap.get(token.address.toLowerCase()) || Number.MAX_SAFE_INTEGER }),
```

File: test/hooks/quoteTokenPriorities.hooks.test.tsx [hunk 10]
```diff
@@ -0,0 +1,147 @@
+import React from 'react'
+import ReactDOM from 'react-dom'
+import { act } from 'react-dom/test-utils'
```
Comment (Velenir) [inline test/hooks/quoteTokenPriorities.hooks.test.tsx:3]: Why suddenly do native react renders, when we use `enzyme` in all other component tests?

File: src/components/TradeWidget/index.tsx [hunk 11]
```diff
@@ -73,17 +76,73 @@ import { updateTradeState } from 'reducers-actions/trade'
...
+        padding: '1rem',
+        height: open ? 'auto' : '3.45rem',
+        minWidth: '29rem',
+        overflow: 'hidden',
+      }}
+    >
+      <a
+        href="#"
+        onClick={handleClose}
+        style={{
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/components/TradeWidget/index.tsx:109]: Gonna remove the whole thing, was mainly for testing

File: src/components/TradeWidget/index.tsx [hunk 12]
```diff
@@ -73,17 +76,73 @@ import { updateTradeState } from 'reducers-actions/trade'
...
+        borderRadius: '0 0 1.5rem 0',
+        color: '#fff',
+        padding: '1rem',
+        height: open ? 'auto' : '3.45rem',
+        minWidth: '29rem',
+        overflow: 'hidden',
+      }}
+    >
+      <a
+        href="#"
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/components/TradeWidget/index.tsx:107]: Thought this was best practice when using non standard anchor tags in react. Nvm

File: test/hooks/quoteTokenPriorities.hooks.test.tsx (hunk 10 above)
Comment (W3stside) [reply test/hooks/quoteTokenPriorities.hooks.test.tsx:3]: For hook testing enzyme works? I honestly just wrote it from memory using act as I'm used to it. Can change if important

File: src/components/TradeWidget/index.tsx [hunk 13]
```diff
@@ -621,20 +626,28 @@ const TradeWidget: React.FC = () => {
...
-            baseToken={receiveToken}
-            quoteToken={sellToken}
-            tabIndex={1}
+            baseToken={baseToken}
+            quoteToken={quoteToken}
+            receiveToken={receiveToken}
+            sellToken={sellToken}
             onSwapPrices={swapPrices}
-            priceShown={priceShown}
+            isPriceInverted={priceShown === 'INVERSE'}
```
Comment (anxolin) [inline src/components/TradeWidget/index.tsx:634]: Refactor isPriceInverted as a const, u have it also in line 649

File: src/components/trade/Price.tsx [hunk 14]
```diff
@@ -225,15 +229,17 @@ export const Price: React.FC<Props> = ({
     inputId: priceInverseInputId,
     precision: DEFAULT_PRECISION,
   })
-  const isDirect = priceShown === 'DIRECT'
+
+  const hideInverse = (!isPriceInverted && wasPriorityAdjusted) || (isPriceInverted && !wasPriorityAdjusted)
```
Comment (anxolin) [inline src/components/trade/Price.tsx:233]: This part is a bit strange, very complex to understand, minimum we need a comment to clarify.
Also, is not true that only one price will be visible. Therefore, only one const should be enought

File: src/components/trade/Price.tsx [hunk 15]
```diff
@@ -225,15 +229,17 @@ export const Price: React.FC<Props> = ({
...
+  const hideDirect = (isPriceInverted && wasPriorityAdjusted) || (!isPriceInverted && !wasPriorityAdjusted)
 
   return (
     <Wrapper>
       <strong>
         Limit Price <OrderBookBtn baseToken={baseToken} quoteToken={quoteToken} />
       </strong>
       {/* using display: none to hide to avoid hook-form reregister */}
-      <PriceInputBox hidden={!isDirect}>
+      <PriceInputBox hidden={hideInverse}>
```
Comment (anxolin) [inline src/components/trade/Price.tsx:242]: This is the direct price, not the inverse, isn't it? why does it say "hide inverse"?

File: src/components/trade/Price.tsx [hunk 16]
```diff
@@ -272,9 +278,9 @@ export const Price: React.FC<Props> = ({
           <SwapPrice
             baseToken={baseToken}
             quoteToken={quoteToken}
-            isPriceInverted
-            showBaseToken
+            isPriceInverted={!wasPriorityAdjusted}
```
Comment (anxolin) [inline src/components/trade/Price.tsx:281]: This make it very difficult to understand, why `isPriceInverted = !wasPriorityAdjusted` 🤯

File: src/components/trade/PriceSuggestions/PriceSuggestionWidget.tsx [hunk 17]
```diff
@@ -19,28 +27,41 @@ export const PriceSuggestionWidget: React.FC<Props> = (props) => {
...
+    sellToken,
     isPriceInverted,
     priceInputId,
     priceInverseInputId,
+    wasPriorityAdjusted,
     onSwapPrices,
   } = props
-  const { id: baseTokenId, decimals: baseTokenDecimals } = baseToken
-  const { id: quoteTokenId, decimals: quoteTokenDecimals } = quoteToken
+  const { id: baseTokenId, decimals: baseTokenDecimals } = receiveToken
```
Comment (anxolin) [inline src/components/trade/PriceSuggestions/PriceSuggestionWidget.tsx:38]: Why this change? is not better to express in terms of base/quote the price suggestions?
Also, I don't think this component should have "wasPriorityAdjusted". This component should know only the base/quote and wether the price has been inverted or not.

File: src/components/trade/PriceSuggestions/PriceSuggestions.tsx [hunk 18]
```diff
@@ -118,7 +117,9 @@ export interface Props {
   // Market
   baseToken: TokenDetails
   quoteToken: TokenDetails
+  sellToken: TokenDetails
   isPriceInverted: boolean
+  wasPriorityAdjusted: boolean
```
Comment (anxolin) [inline src/components/trade/PriceSuggestions/PriceSuggestions.tsx:122]: Please, let's not propagate `wasPriorityAdjusted` this prop down all the components. you just let the component know what's your selection of base/quote and if it is inverted or not.

**Comment by anxolin** [comment]
Functionally I believe it works as expected.
There's still some complexities added that can be simplified imo.

File: src/components/trade/Price.tsx (hunk 14 above)
Comment (W3stside) [reply src/components/trade/Price.tsx:233]: fair enough on 1 constant required

Comment (W3stside) [reply src/components/trade/Price.tsx:233]: I am going to rework thru the other comments to try and cut complexity but i still think the 2nd flag is required to balance the inverse/direct price with the act of changing the token order when getting pricing. Will add comment reflecting this

File: src/components/trade/Price.tsx (hunk 15 above)
Comment (W3stside) [reply src/components/trade/Price.tsx:242]: yup

**Comment by W3stside** [comment]
closing
